### PR TITLE
fix(lint): windows compatibility, collision of lib names

### LINKED ISSUE #285: @nrwl/schematics@0.7.4 breaks calculation of relative paths in nx-enforce-module-boundaries on windows
nx-enforce-module-boundaries produces error: "...library imports must start..."
for cases like "import { AppComponent } from './app.component';" in app.component.spec.ts
Both files are located in the same directory.
The problem is not reproduced in @nrwl/schematics@0.7.3
OS: Windows

### PR DESCRIPTION
**Pull request by skydever**
modify windows path to be comparable with import statement

libraryRoot/lazyLoaded: consider lib names that start with name of other lib
(eg: my-shared, my-shared-extended)

Closes #285 #303

### CODE REVIEW

**Comment by skydever** [comment]
at [my fork](https://github.com/skydever/nx/pull/5) the ci was passing ...

**Comment by skydever** [comment]
started to adopt the rule tests. my test workspace should cover a lot of usecases, so I thought of using it as the basic options for a set of tests. started using it as runRule defaults, moved content to 1st param - justing running through different import statements/content. it is a windows setup (win paths), I can also add a config for linux paths (just switching the defaults - two methods maybe, runRuleWin, runRuleLinux or like that ...). just wanted to let you know ...

**Comment by skydever** [comment]
@vsavkin I think I am ready for a review now. The last refactoring of my patch also made the test adoption easier than I thought at the beginning, hope it is a way to go.

**Comment by ThomasBurleson** [comment]
@skydever - please squash your commits and rebase w/ latest code from HEAD.

**Comment by ThomasBurleson** [comment]
@thelgevold - can you review these changes ?

**Comment by skydever** [comment]
@ThomasBurleson is it OK like this (squash, rebase)?

**Comment by skydever** [comment]
...just fixed [this line](https://github.com/nrwl/nx/pull/290/files#diff-d0d9d1f9a4bf00b6435695bad3d15e68R9) at the spec from
```
{ allow: ['@mycompany/mylib/deep'], lazy: ['@mycompany/myliblazy'] }
```
to
```
{ allow: ['@mycompany/mylib/deep'], lazyLoad: ['myliblazy'] }
```

**Comment by skydever** [comment]
...started with some refactoring, would look like [this](https://github.com/skydever/nx/pull/6). what do you think about it?

**Comment by ThomasBurleson** [comment]
@skydever - squashing looks good. Waiting for @vsavkin  review.

File: packages/schematics/src/tslint/nxEnforceModuleBoundariesRule.spec.ts [hunk 1]
```diff
@@ -6,7 +6,7 @@ import { Rule } from './nxEnforceModuleBoundariesRule';
 describe('Enforce Module Boundaries', () => {
   it('should not error when everything is in order', () => {
     const failures = runRule(
-      { allow: ['@mycompany/mylib/deep'] },
+      { allow: ['@mycompany/mylib/deep'], lazyLoad: ['myliblazy'] },
```
Comment (vsavkin) [inline packages/schematics/src/tslint/nxEnforceModuleBoundariesRule.spec.ts]: Why is this addition required? It doesn't look like we reference myliblazy anywhere in tests.

File: packages/schematics/src/tslint/nxEnforceModuleBoundariesRule.ts [hunk 2]
```diff
@@ -55,7 +55,13 @@ class EnforceModuleBoundariesWalker extends Lint.RuleWalker {
...
   ) {
     super(sourceFile, options);
-    this.roots = [...roots].sort((a, b) => a.length - b.length);
+    //
+    // TODO
+    // remove sort completely in favor of appending '/' at libraryRoot()?
+    // sort logic does not work for lazyLoaded, so one logic to check for name collision?
+    //
+    // fix: sort from longest to shortest
+    this.roots = [...roots].sort((a, b) => b.length - a.length);
```
Comment (vsavkin) [inline packages/schematics/src/tslint/nxEnforceModuleBoundariesRule.ts]: could you add a test for this change?

File: packages/schematics/src/tslint/nxEnforceModuleBoundariesRule.ts [hunk 3]
```diff
@@ -55,7 +55,13 @@ class EnforceModuleBoundariesWalker extends Lint.RuleWalker {
     private roots: string[]
   ) {
     super(sourceFile, options);
-    this.roots = [...roots].sort((a, b) => a.length - b.length);
+    //
+    // TODO
+    // remove sort completely in favor of appending '/' at libraryRoot()?
+    // sort logic does not work for lazyLoaded, so one logic to check for name collision?
```
Comment (vsavkin) [inline packages/schematics/src/tslint/nxEnforceModuleBoundariesRule.ts]: Why doesn't sort logic work for lazyLoaded? Could you expand on this?

**Review by vsavkin — CHANGES_REQUESTED** [review]

**Comment by vsavkin** [comment]
Left some comments

File: packages/schematics/src/tslint/nxEnforceModuleBoundariesRule.ts (hunk 2 above)
Comment (skydever) [reply packages/schematics/src/tslint/nxEnforceModuleBoundariesRule.ts]: I can add a test case where I have the libs `mylib` and `mylibtwo`. I "create" a file at `/dev/workspace/libs/mylibtwo/mycomp.ts` and do a relative import like `./mycomptwo`. this should pass and not error with `library imports must start with @myworkspace/`. did you mean something like that?

the test `should not error when relatively importing the same library` did pass prior to this pr because the file was created in the shorter named lib and the shorter named lib came first because of the wrong sort ...

besides all of that, I did some refactoring concerning the lint rule (and added glob support for the `allow` option). do you mind having a look [here](https://github.com/skydever/nx/pull/7)? Is it good/bad? The concept was to evaluate the `root` of both source and target at the beginning, then the checks seemed easier...

**Review by skydever — COMMENTED** [review]
still some questions. thx for the review!

**Comment by skydever** [comment]
@vsavkin added the changes, will rebase + squash again (sry for duplicate review) ...

**Comment by vsavkin** [comment]
Thank you!
